### PR TITLE
delete: fix glusterBlockDeleteRemoteAsync ret value

### PR DESCRIPTION
**Pull request by pkalever**
Before:
[root@localhost gluster-block]# gluster-block delete sample/block
FAILED ON:   192.168.124.217
SUCCESSFUL ON:   192.168.124.6 192.168.124.163
RESULT: SUCCESS

After:
[root@localhost gluster-block]# gluster-block delete sample/block
FAILED ON:   192.168.124.217
SUCCESSFUL ON:   192.168.124.163 192.168.124.6
RESULT: FAIL

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

### CODE REVIEW

**Comment by pkalever** [comment]
@pranithk @lxbsz please review

**Comment by pkalever** [comment]
@pranithk as you are aware,


We have covered case:
-----------------------

tree 04238f51403458e3fcf4906ef56e7fc425e44b95                                    
parent 0b0b7db14159e009d60435c38ded42a5606472ad                                  
author Prasanna Kumar Kalever <prasanna.kalever@redhat.com> Tue Jun 5 13:39:31 2018 +0530
committer Pranith Kumar Karampuri <pranith.k@gmail.com> Tue Jun 5 19:35:58 2018 +0800
                                                                                 
delete: defend on block not loaded and CLEANUPFAIL status                        
                                                                                 
Consider the following case:                                                     
                                                                                 
* When gluster-block delete was executed for the first time, on this node,       
  deletion was successful. But before it can send the response,                  
  gluster-blockd died (addr meta status would be CLEANUPFAIL)                    
                                                                                 
* When the user attempts gluster-block deletion, it should succeed.              
                                                                                                                                                                                                                   
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>              


With above change, if the block is not loaded on 1 node out of 3, the first delete will fail saying block not loaded in the logs and the second delete will succeed. Are we okay with this ?

**Comment by pranithk** [comment]
@pkalever Please add the problem statement and what you are trying to fix.

**Comment by pranithk** [comment]
> With above change, if the block is not loaded on 1 node out of 3, the first delete will fail saying block not loaded in the logs and the second delete will succeed. Are we okay with this ?

@pkalever It is not a good user experience to fail and then succeed.

**Comment by pkalever** [comment]
@pranithk @lxbsz please review.

Issue https://github.com/gluster/gluster-block/issues/96 is to track the Hints fix, for now will just log.

File: rpc/block_svc_routines.c [hunk 1]
```diff
@@ -258,6 +260,67 @@ blockStr2arrayAddToJsonObj(json_object *json_obj, char *string, char *label)
...
+    LOG("mgmt", GB_LOG_ERROR, "%s", reply->out);
+    goto out;
+  } else if (ret == 1) {
+    is_loaded = false;
+    GB_ASPRINTF(&reply->out, "Block '%s' may be not loaded\n", block_name);
+    LOG("mgmt", GB_LOG_ERROR, "%s", reply->out);
+  }
+  GB_FREE(exec);
+
+  if (GB_ASPRINTF(&exec, GB_SAVECONFIG_CHECK, block_name) == -1) {
```
Comment (pranithk) [inline rpc/block_svc_routines.c]: Please set return value indicating error

File: rpc/block_svc_routines.c [hunk 2]
```diff
@@ -258,6 +260,67 @@ blockStr2arrayAddToJsonObj(json_object *json_obj, char *string, char *label)
...
+  if (GB_ASPRINTF(&exec, GB_TGCLI_CHECK, block_name, gbid) == -1) {
+    goto out;
+  }
+
+  ret = gbRunner(exec);
+  if (ret == -1) {
+    GB_ASPRINTF(&reply->out, "command exit abnormally for %s\n", block_name);
+    LOG("mgmt", GB_LOG_ERROR, "%s", reply->out);
+    goto out;
+  } else if (ret == 1) {
```
Comment (pranithk) [inline rpc/block_svc_routines.c]: exit value 0 means success and non-zero is failure. So are we sure that ret value is going to be 1 for this failure? Same comment for the next set of if/else as well

File: rpc/block_svc_routines.c [hunk 3]
```diff
@@ -258,6 +260,67 @@ blockStr2arrayAddToJsonObj(json_object *json_obj, char *string, char *label)
...
+    goto out;
+  }
+
+  ret = gbRunner(exec);
+  if (ret == -1) {
+    GB_FREE(reply->out);
+    GB_ASPRINTF(&reply->out, "command exit abnormally for %s\n", block_name);
+    LOG("mgmt", GB_LOG_ERROR, "%s", reply->out);
+    goto out;
+  } else if (ret == 1 || ret == 2) {  /* search not found or savefile not exist */
```
Comment (pranithk) [inline rpc/block_svc_routines.c]: same comment as above

File: rpc/block_svc_routines.c (hunk 1 above)
Comment (pkalever) [reply rpc/block_svc_routines.c]: Good catch.

File: rpc/block_svc_routines.c (hunk 2 above)
Comment (pkalever) [reply rpc/block_svc_routines.c]: Yes, it will be 1 (not something newly introduced)

File: rpc/block_svc_routines.c (hunk 3 above)
Comment (pkalever) [reply rpc/block_svc_routines.c]: Yes, 2 is extra here for file not found case (grep on savefile case)

**Comment by pkalever** [comment]
@pranithk update the ret = -1 on GB_ASPRINTF() fail

Thanks!

File: rpc/block_svc_routines.c (hunk 3 above)
Comment (pranithk) [reply rpc/block_svc_routines.c]: Isn't it better to flip the condition? If ret == 0 consider success otherwise consider failure? Isn't that more robust?

**Comment by pkalever** [comment]
Added:
 /* if block is loaded, skip the rest */
+  if (!ret) {
+    goto out;
+  }

**Comment by pkalever** [comment]
Added:

```
if (list->nhosts != info->mpath) {                                      
  errCode = ENOENT;                                                     
                                                                        
  GB_ASPRINTF(&errMsg, "Some of the nodes are missing configuration, " \
              "please look at logs for recent operations on the %d.",   
              blk->block_name);                                         
  LOG("mgmt", GB_LOG_ERROR, "%s", errMsg);                              
                                                                        
  goto out;                                                             
}                                                                       

```

Case: when delete is issued, in case if there are 2 nodes where delete succeeded and on 1 node, block not loaded. It was returning success.

File: rpc/block_svc_routines.c (hunk 3 above)
Comment (pkalever) [reply rpc/block_svc_routines.c]: will generalize this part later, may be in a different revision as this is well tested lets live with it for now :-)

**Comment by pkalever** [comment]
I have tested this patch PR with various cases like:

1. block not loaded, but exist in the saveconfig.json
result: delete and modify both fail
2. block loaded, but not exist in the saveconfig.json
result: delete and modify both success
3. block loaded, but saveconfig.json file missing
result: delete and modify both success
4. block not loaded and not exist in the saveconfig.json
result: delete success and modify fail
5. block deleted on one node, then the subsequent modify will error out with 
"errMsg":"Some of the nodes are missing configuration, please look at logs for recent operations on the 'block1'."

Thanks!

**Comment by pkalever** [comment]
@pranithk tags refreshed.

**Pull request merged**